web/shared/src/internal_url.ts

```diff
@@ -67,3 +67,60 @@ export function by_stream_topic_url(
         maybe_get_stream_name,
     )}/topic/${encodeHashComponent(topic)}`;
 }
+/**
+ *
+ * @param url_str string containing the url
+ * @param maybe_get_stream_name function to get stream name from id
+ * @returns null if the url is malformed or belongs to a different server
+ *         [stream_id, topic] if the url is a valid narrow url. `topic` may
+ *          be undefined, if absent from the url.
+ */
+export function stream_topic_hash_from_url(
+    url_str: string,
+    maybe_get_stream_name: MaybeGetStreamName,
+): string[] {
+    let url;
+    try {
+        url = new URL(url_str);
+    } catch {
+        // url is invalid
+        return [];
+    }
+    // eslint-disable-next-line no-undef
+    if (url.origin !== window.location.origin) {
+        // not a url of the same zulip server
+        return [];
+    }
```

Comment:
How does this block handle a relative link, i.e. one starting with `/#narrow/...`?

Probably it'll be OK/correct to convert those too, even though we don't generate them usually.

Reply:
As discussed on CZO, relative links won't be transformed.